File: exo5_8.c
#include <stdio.h>
#include <stdlib.h>
#include <pthread.h>
#include <semaphore.h>
#include <unistd.h>

struct RDV_2 {
    sem_t *sem_t1; // Signalé par T1
    sem_t *sem_t2; // Signalé par T2
};

void* thread_T1(void* arg) {
    struct RDV_2 *ctx = (struct RDV_2*)arg;
    for(int i = 1; i <= 10; i++) {
        sleep(2); // Traitement fixe
        printf("[T1] Itération %d : Arrivé au rendez-vous.\n", i);
        
        sem_post(ctx->sem_t1);
        sem_wait(ctx->sem_t2);
        
        printf("[T1] Itération %d : Rendez-vous réussi, je continue.\n", i);
    }
    pthread_exit(NULL);
}

void* thread_T2(void* arg) {
    struct RDV_2 *ctx = (struct RDV_2*)arg;
    for(int i = 1; i <= 10; i++) {
        sleep(rand() % 6 + 4); 
        printf("[T2] Itération %d : Arrivé au rendez-vous.\n", i);
        
        sem_post(ctx->sem_t2);
        sem_wait(ctx->sem_t1); 
        
        printf("[T2] Itération %d : Rendez-vous réussi, je continue.\n", i);
    }
    pthread_exit(NULL);
}

int main() {
    pthread_t t1, t2;
    sem_t s1, s2;
    sem_init(&s1, 0, 0);
    sem_init(&s2, 0, 0);

    struct RDV_2 *ctx = malloc(sizeof(struct RDV_2));
    ctx->sem_t1 = &s1; ctx->sem_t2 = &s2;

    pthread_create(&t1, NULL, thread_T1, ctx);
    pthread_create(&t2, NULL, thread_T2, ctx);

    pthread_join(t1, NULL); pthread_join(t2, NULL);

    sem_destroy(&s1); sem_destroy(&s2);
    free(ctx);
    return 0;
}

File: exo5_8.py
import threading
import time
import random
sem1 = threading.Semaphore(0)
sem2 = threading.Semaphore(0)

def thread_t1():
    for i in range(1, 11):
        print(f"[T1] Itération {i} : Traitement (2s)...")
        time.sleep(2)
        
        print(f"[T1] Arrivé au rendez-vous {i}. Attend T2...")
        sem1.release() 
        sem2.acquire()
        print(f"[T1] Rendez-vous {i} terminé. On continue.\n")

def thread_t2():
    for i in range(1, 11):
        duree = random.randint(4, 9)
        print(f"[T2] Itération {i} : Traitement ({duree}s)...")
        time.sleep(duree)
        
        print(f"[T2] Arrivé au rendez-vous {i}. Attend T1...")
        sem2.release()
        sem1.acquire()
        print(f"[T2] Rendez-vous {i} terminé. On continue.\n")

def main_a():
    t1 = threading.Thread(target=thread_t1)
    t2 = threading.Thread(target=thread_t2)
    t1.start()
    t2.start()
    t1.join()
    t2.join()

if __name__ == "__main__":
    main_a()
